File: ds_algo/grind75/110_BalancedBinaryTree.py
### GRIND 75 SECOND PASS ###
"""
approach:
    recursive:
        ask left and right child nodes about their height
        compare heights: good to go if their heights differ by < 2
        ask left and right if their children *also* were balanced (this will avoid hidden imbalances)
        return True if these three cases are true
        base case: return True if root is None
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# recursive
# T: O(n)
# S: O(n) for recursive space
class Solution:   
    class Height:
        def __init__(self):
            self.height = 0
    def isBalanced(self, root: Optional[TreeNode], height: Height=None) -> bool:
        if root is None:
            return True
        # heights
        leftHeight = self.Height()
        rightHeight = self.Height()
        
        # recurse on left and right children (aka ask if they are balanced) -> bool
        leftBalanced = self.isBalanced(root.left, leftHeight)
        rightBalanced = self.isBalanced(root.right, rightHeight)
        
        # compute the height of the current node from the children
        if height: # height was passed in by reference on recursion (=None on first function call)
            height.height = 1 + max(leftHeight.height, rightHeight.height)
        
        # we want to check two things here:
        # 1) are the heights of my left and right child's subtrees within 1 of each other?
        # 2) was the same true for their children? (i.e. avoid hidden imbalances)
        if abs(leftHeight.height - rightHeight.height) < 2:
            return (leftBalanced and rightBalanced) # will be True if left and right are balanced + their children are balanced
        
        # if we haven't returned by now, the tree is not balanced
        return False

    
# post-order traversal (found this one in the forum)
# T: O(n) number of nodes in the tree
# S: O(?) I'm inclined to say O(n) space complexity because of the depths dict
class Solution(object):
    def isBalanced(self, root):
        stack, node, last, depths = [], root, None, {}
        while stack or node:
            if node:
                stack.append(node)
                node = node.left
            else: # if the node we examined on the last iteration was a leaf, then its children are None
                node = stack[-1]
                if not node.right or last == node.right:
                    node = stack.pop()
                    left, right  = depths.get(node.left, 0), depths.get(node.right, 0)
                    if abs(left - right) > 1: return False
                    depths[node] = 1 + max(left, right)
                    last = node
                    node = None
                else:
                    node = node.right
        return True

File: ds_algo/grind75/110_BalancedBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

/*
approach:

*/

class Solution {
 public:
  bool isBalanced(TreeNode* root, int* height = nullptr) {
    // define right and left tree heights to be referenced in recursive calls
    int leftHeight = 0;
    int rightHeight = 0;

    // check if we are at a leaf node
    // if so, return true (this will help us establish height later on)
    if (root == nullptr) {
      return true;
    }

    // recurse to the right and left branches
    // this will allow us to mark if the branches are balanced within themselves(?)
    int leftBool = isBalanced(root->left, &leftHeight);
    int rightBool = isBalanced(root->right, &rightHeight);

    // keep track of the total (largest) height on this branch
    // it's important to set this *after* the recursive call
    // because the recursive call incremented this call's leftHeight and rightHeight
    if (height) {
      *height = max(leftHeight, rightHeight) + 1;
    }

    // if the current heights don't differ by > 1, return true only if both branches concur
    if (abs(leftHeight - rightHeight) < 2) {
      return leftBool && rightBool;
    }

    return false;
  }
};